cilium-cli/connectivity/sniff/sniffer.go

```diff
@@ -25,17 +25,122 @@ const (
 	// ModeSanity: expect to observe packets matching the filter, to be
 	// leveraged as a sanity check to verify that the filter is correct.
 	ModeSanity Mode = "sanity"
+
+	// Max wait time for tcpdump to start/stop in remote shell.
+	sniffScriptTimeout = 10 * time.Second
+	// Max timeout before aborting k8s connection in Start/Stop; must exceed sniffScriptTimeout.
+	sniffConnectionTimeout = sniffScriptTimeout + 5*time.Second
+	// Max remote sniffer runtime to prevent lingering processes in the pod.
+	// NOTE: too low may kill tcpdump while test is running.
+	sniffKillTimeout = sniffConnectionTimeout * 4
+
+	// Command executed to start the remote tcpdump in background inside a pod.
+	//
+	// We send tcpdump output to a file inside the pod rather than stdout,
+	// to avoid GH Issue #38643: tcpdump failing with error 14 due to one of
+	// its output streams not being available anymore.
+	// (tcpdump: Unable to write output: Broken pipe)
+	//
+	// Runs silently: succeeds if tcpdump is active ("listening on <iface>"),
+	// else fails with its error code. Includes `timeout 60` to prevent orphaned
+	// tcpdump if sniffer isn't properly stopped.
+	sniffScriptStartTmpl = `
+{
+	# Cleanup any leftover files from previous runs.
+	rm -f {{ .PidPath }} {{ .DumpPath }} {{ .LogPath }}
+
+	# Start tcpdump directly in background and capture its PID.
+	timeout {{ .KillSeconds }} tcpdump -i {{ .Iface }} -c {{ .PktCount }} --immediate-mode -U -w {{ .DumpPath }} "{{ .Filter }}" > {{ .LogPath }} 2>&1 &
+	pid=$!
+	echo $pid > {{ .PidPath }}
+
+	# Wait for startup confirmation and exit with success.
+	# Exit with error code if process is not running.
+	for i in $(seq 1 {{ .WaitSeconds }}); do
+		sleep 1
+		grep -q "listening on {{ .Iface }}" {{ .LogPath }} && exit 0
+		kill -0 $pid 2>/dev/null || {
+			wait $pid
+			exit $?
+		}
+	done
+
+	# Process is running but did not receive confirmation, kill and exit.
+	kill -9 $pid 2>/dev/null
+	wait $pid
+	exit $?
+}
+`
+
+	// Command executed to stop the remote tcpdump inside a pod.
+	// On successful sniffer stop, prints captured packets and their total count to stdout.
+	sniffScriptStopTmpl = `
+{
+	# Print captured packets and their count to stdout.
+	# Redirect stderr to /dev/null, as tcpdump logs to stderr.
+	report() {
+		tcpdump -n -r {{ .DumpPath }} --count 2>/dev/null || true
+		tcpdump -n -r {{ .DumpPath }}         2>/dev/null || true
+	}
+
+	# Ignore signals received when killing tcpdump.
+	trap '' TERM INT HUP
+
+	# Exit with error if tcpdump is not running.
+	if [ ! -f "{{ .PidPath }}" ]; then
+		exit 199
+	fi
+
+	pid=$(cat {{ .PidPath }} )
+
+	# Kill the process if it is still running.
+	# Return an error if tcpdump timed out before capturing any packets.
+	if kill -0 $pid 2>/dev/null; then
+		kill -2 $pid 2>/dev/null
+	elif [ "$(tcpdump -r {{ .DumpPath }} --count 2>/dev/null)" = "0 packets" ]; then
```

Comment:
@gentoo-root I believe this is incorrect for the ModeAssert where having `0 packets` is good.
Isn't this checked later in Go anyway?

Reply:
Nope, given we now have also `-c 1000` for ModeAssert.
Tcpdump might have already recorded 1000 packets and exited before our check here.